// @waves/signature-generator/src/utils/convert.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
const converters_1 = require("../libs/converters");
const data_entities_1 = require("@waves/data-entities");
function performBitwiseAnd(a, b) {
    const sa = a.toString(2).split('.')[0];
    const sb = b.toString(2).split('.')[0];
    const len = Math.min(sa.length, sb.length);
    const s1 = sa.slice(sa.length - len);
    const s2 = sb.slice(sb.length - len);
    let result = new Array(len);
    for (let i = len - 1; i >= 0; i--) {
        result[i] = (s1[i] === '1' && s2[i] === '1') ? '1' : '0';
    }
    return parseInt(result.join(''), 2);
}
exports.default = {
    booleanToBytes(input) {
        if (typeof input !== 'boolean') {
            throw new Error('Boolean input is expected');
        }
        return input ? [1] : [0];
    },
    shortToByteArray(input) {
        if (typeof input !== 'number') {
            throw new Error('Numeric input is expected');
        }
        return converters_1.default.int16ToBytes(input, true);
    },
    bytesToByteArrayWithSize(input) {
        if (!(input instanceof Array || input instanceof Uint8Array)) {
            throw new Error('Byte array or Uint8Array input is expected');
        }
        else if (input instanceof Array && !(input.every((n) => typeof n === 'number'))) {
            throw new Error('Byte array contains non-numeric elements');
        }
        if (!(input instanceof Array)) {
            input = Array.prototype.slice.call(input);
        }
        const lengthBytes = converters_1.default.int16ToBytes(input.length, true);
        return [...lengthBytes, ...input];
    },
    longToByteArray(input) {
        if (typeof input !== 'number') {
            throw new Error('Numeric input is expected');
        }
        const bytes = new Array(7);
        for (let k = 7; k >= 0; k--) {
            bytes[k] = input & (255);
            input = input / 256;
        }
        return bytes;
    },
    bigNumberToByteArray(input) {
        if (!(input instanceof data_entities_1.BigNumber)) {
            throw new Error('BigNumber input is expected');
        }
        const performBitwiseAnd255 = performBitwiseAnd.bind(null, new data_entities_1.BigNumber(255));
        const bytes = new Array(7);
        for (let k = 7; k >= 0; k--) {
            bytes[k] = performBitwiseAnd255(input);
            input = input.div(256);
        }
        return bytes;
    },
    stringToByteArray(input) {
        if (typeof input === 'number') {
            input = String(input);
        }
        if (typeof input !== 'string') {
            throw new Error('String input is expected');
        }
        return converters_1.default.stringToByteArray(input);
    },
    stringToByteArrayWithSize(input) {
        if (typeof input === 'number') {
            input = String(input);
        }
        if (typeof input !== 'string') {
            throw new Error('String input is expected');
        }
        const stringBytes = converters_1.default.stringToByteArray(input);
        const lengthBytes = converters_1.default.int16ToBytes(stringBytes.length, true);
        return [...lengthBytes, ...stringBytes];
    }
};
