// @waves/signature-adapter/dist/Signable.js
"use strict";
var __assign = (this && this.__assign) || Object.assign || function(t) {
    for (var s, i = 1, n = arguments.length; i < n; i++) {
        s = arguments[i];
        for (var p in s) if (Object.prototype.hasOwnProperty.call(s, p))
            t[p] = s[p];
    }
    return t;
};
var __awaiter = (this && this.__awaiter) || function (thisArg, _arguments, P, generator) {
    return new (P || (P = Promise))(function (resolve, reject) {
        function fulfilled(value) { try { step(generator.next(value)); } catch (e) { reject(e); } }
        function rejected(value) { try { step(generator["throw"](value)); } catch (e) { reject(e); } }
        function step(result) { result.done ? resolve(result.value) : new P(function (resolve) { resolve(result.value); }).then(fulfilled, rejected); }
        step((generator = generator.apply(thisArg, _arguments || [])).next());
    });
};
var __generator = (this && this.__generator) || function (thisArg, body) {
    var _ = { label: 0, sent: function() { if (t[0] & 1) throw t[1]; return t[1]; }, trys: [], ops: [] }, f, y, t, g;
    return g = { next: verb(0), "throw": verb(1), "return": verb(2) }, typeof Symbol === "function" && (g[Symbol.iterator] = function() { return this; }), g;
    function verb(n) { return function (v) { return step([n, v]); }; }
    function step(op) {
        if (f) throw new TypeError("Generator is already executing.");
        while (_) try {
            if (f = 1, y && (t = op[0] & 2 ? y["return"] : op[0] ? y["throw"] || ((t = y["return"]) && t.call(y), 0) : y.next) && !(t = t.call(y, op[1])).done) return t;
            if (y = 0, t) op = [op[0] & 2, t.value];
            switch (op[0]) {
                case 0: case 1: t = op; break;
                case 4: _.label++; return { value: op[1], done: false };
                case 5: _.label++; y = op[1]; op = [0]; continue;
                case 7: op = _.ops.pop(); _.trys.pop(); continue;
                default:
                    if (!(t = _.trys, t = t.length > 0 && t[t.length - 1]) && (op[0] === 6 || op[0] === 2)) { _ = 0; continue; }
                    if (op[0] === 3 && (!t || (op[1] > t[0] && op[1] < t[3]))) { _.label = op[1]; break; }
                    if (op[0] === 6 && _.label < t[1]) { _.label = t[1]; t = op; break; }
                    if (t && _.label < t[2]) { _.label = t[2]; _.ops.push(op); break; }
                    if (t[2]) _.ops.pop();
                    _.trys.pop(); continue;
            }
            op = body.call(thisArg, _);
        } catch (e) { op = [6, e]; y = 0; } finally { f = t = 0; }
        if (op[0] & 5) throw op[1]; return { value: op[0] ? op[1] : void 0, done: true };
    }
};
Object.defineProperty(exports, "__esModule", { value: true });
var prepareTx_1 = require("./prepareTx");
var signature_generator_1 = require("@waves/signature-generator");
var Signable = /** @class */ (function () {
    function Signable(forSign, adapter) {
        var _this = this;
        this._proofs = [];
        this._forSign = __assign({}, forSign);
        this._adapter = adapter;
        this._prepare = prepareTx_1.getSchemaByType(forSign.type);
        if (!this._forSign.data.timestamp) {
            this._forSign.data.timestamp = Date.now();
        }
        if (this._forSign.data.proofs) {
            this._proofs = this._forSign.data.proofs.slice();
        }
        if (!this._prepare) {
            if (forSign.type !== prepareTx_1.SIGN_TYPE.CUSTOM) {
                throw new Error("Can't find prepare api for tx type \"" + forSign.type + "\"!");
            }
            else {
                this._prepare = {
                    sign: forSign.signProcessor || (function (data) { return data; }),
                    api: forSign.apiProcessor || (function (data) { return data; })
                };
            }
        }
        var generator;
        if (forSign.type === prepareTx_1.SIGN_TYPE.CUSTOM) {
            generator = forSign.generator;
            this._signMethod = 'signRequest';
        }
        else {
            generator = prepareTx_1.SIGN_TYPES[forSign.type].signatureGenerator;
            this._signMethod = prepareTx_1.SIGN_TYPES[forSign.type].adapter;
        }
        if (!generator) {
            throw new Error("Unknown data type " + forSign.type + "!");
        }
        this._prepare.sign(forSign.data, true);
        this._bytePromise = Promise.all([
            this._adapter.getPublicKey(),
            this._adapter.getAddress()
        ]).then(function (_a) {
            var senderPublicKey = _a[0], sender = _a[1];
            var dataForSign = _this._prepare.sign(__assign({ sender: sender, senderPublicKey: senderPublicKey }, forSign.data));
            return new generator(dataForSign).getBytes();
        });
    }
    Signable.prototype.getTxData = function () {
        return __assign({}, this._forSign.data);
    };
    Signable.prototype.getSignData = function () {
        return __awaiter(this, void 0, void 0, function () {
            var senderPublicKey, sender;
            return __generator(this, function (_a) {
                switch (_a.label) {
                    case 0: return [4 /*yield*/, this._adapter.getPublicKey()];
                    case 1:
                        senderPublicKey = _a.sent();
                        return [4 /*yield*/, this._adapter.getAddress()];
                    case 2:
                        sender = _a.sent();
                        return [2 /*return*/, this._prepare.sign(__assign({ sender: sender, senderPublicKey: senderPublicKey }, this._forSign.data))];
                }
            });
        });
    };
    Signable.prototype.sign2fa = function (options) {
        var _this = this;
        var code = options.code;
        return this._adapter.getAddress()
            .then(function (address) {
            return options.request({
                address: address,
                code: code,
                signData: _this._forSign
            });
        })
            .then(function (signature) {
            _this._proofs.push(signature);
            return _this;
        });
    };
    Signable.prototype.addProof = function (signature) {
        if (this._proofs.indexOf(signature) !== -1) {
            this._proofs.push(signature);
        }
        return this;
    };
    Signable.prototype.getId = function () {
        return this._bytePromise.then(signature_generator_1.utils.crypto.buildTransactionId);
    };
    Signable.prototype.sign = function () {
        var _this = this;
        this._makeSignPromise();
        return this._signPromise.then(function () { return _this; });
    };
    Signable.prototype.getSignature = function () {
        this._makeSignPromise();
        return this._signPromise;
    };
    Signable.prototype.getBytes = function () {
        return this._bytePromise;
    };
    Signable.prototype.getMyProofs = function () {
        var _this = this;
        return Promise.all([
            this.getBytes(),
            this._adapter.getPublicKey()
        ]).then(function (_a) {
            var bytes = _a[0], publicKey = _a[1];
            return _this._proofs.filter(function (signature) {
                try {
                    return signature_generator_1.utils.crypto.isValidSignature(bytes, signature, publicKey);
                }
                catch (e) {
                    return false;
                }
            });
        });
    };
    Signable.prototype.hasMySignature = function () {
        return this.getMyProofs().then(function (proofs) { return !!proofs.length; });
    };
    Signable.prototype.addMyProof = function () {
        var _this = this;
        return this.hasMySignature().then(function (hasMySignature) {
            if (!hasMySignature) {
                return _this.getSignature().then(function (signature) {
                    _this._proofs.push(signature);
                    return signature;
                });
            }
            else {
                return _this.getMyProofs().then(function (list) { return list[list.length - 1]; });
            }
        });
    };
    Signable.prototype.getDataForApi = function () {
        var _this = this;
        return Promise.all([
            this._adapter.getPublicKey(),
            this._adapter.getAddress(),
            this.addMyProof()
        ]).then(function (_a) {
            var senderPublicKey = _a[0], sender = _a[1];
            var proofs = _this._proofs.slice();
            return _this._prepare.api(__assign({ senderPublicKey: senderPublicKey, sender: sender }, _this._forSign.data, { proofs: proofs }));
        });
    };
    Signable.prototype._makeSignPromise = function () {
        var _this = this;
        if (!this._signPromise) {
            this._signPromise = this._bytePromise.then(function (bytes) {
                return _this._adapter[_this._signMethod](bytes, _this._getAmountPrecision(), _this._forSign);
            });
            this._signPromise.catch(function () {
                _this._signPromise = null;
            });
        }
        return this;
    };
    Signable.prototype._getAmountPrecision = function () {
        var data = this._forSign.data;
        return data.amount && data.amount.asset && data.amount.asset.precision ? data.amount.asset.precision : 0;
    };
    return Signable;
}());
exports.Signable = Signable;
