// waves-transactions/transactions/data.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
const waves_crypto_1 = require("waves-crypto");
const transactions_1 = require("../transactions");
const generic_1 = require("../generic");
const marshall_1 = require("@waves/marshall");
const typeMap = {
    integer: ['integer', 0, waves_crypto_1.LONG],
    number: ['integer', 0, waves_crypto_1.LONG],
    boolean: ['boolean', 1, waves_crypto_1.BYTE],
    string: ['string', 3, waves_crypto_1.LEN(waves_crypto_1.SHORT)(waves_crypto_1.STRING)],
    binary: ['binary', 2, (s) => waves_crypto_1.LEN(waves_crypto_1.SHORT)(waves_crypto_1.BASE64_STRING)(s.slice(7))],
    _: ['binary', 2, waves_crypto_1.LEN(waves_crypto_1.SHORT)(waves_crypto_1.BYTES)],
};
const mapType = (value) => typeMap[typeof value] || typeMap['_'];
function data(paramsOrTx, seed) {
    const type = transactions_1.TRANSACTION_TYPE.DATA;
    const version = paramsOrTx.version || 1;
    const seedsAndIndexes = generic_1.convertToPairs(seed);
    const senderPublicKey = generic_1.getSenderPublicKey(seedsAndIndexes, paramsOrTx);
    if (!Array.isArray(paramsOrTx.data))
        throw new Error('["data should be array"]');
    const _timestamp = paramsOrTx.timestamp || Date.now();
    let bytes = waves_crypto_1.concat(waves_crypto_1.BYTE(transactions_1.TRANSACTION_TYPE.DATA), waves_crypto_1.BYTE(1), waves_crypto_1.BASE58_STRING(senderPublicKey), waves_crypto_1.COUNT(waves_crypto_1.SHORT)((x) => waves_crypto_1.concat(waves_crypto_1.LEN(waves_crypto_1.SHORT)(waves_crypto_1.STRING)(x.key), [mapType(x.value)[1]], mapType(x.value)[2](x.value)))(paramsOrTx.data), waves_crypto_1.LONG(_timestamp));
    const computedFee = (Math.floor(1 + (bytes.length + 8 /*feeLong*/ - 1) / 1024) * 100000);
    const tx = {
        type,
        version,
        senderPublicKey,
        fee: computedFee,
        timestamp: _timestamp,
        proofs: paramsOrTx.proofs || [],
        id: '',
        data: paramsOrTx.data && paramsOrTx.data.map((x) => {
            if (x.type)
                return x;
            else {
                const type = mapType(x.value)[0];
                return {
                    type,
                    key: x.key,
                    value: type === 'binary' ? 'base64:' + Buffer.from(x.value).toString('base64') : x.value,
                };
            }
        }),
    };
    const bytes1 = marshall_1.binary.serializeTx(tx);
    seedsAndIndexes.forEach(([s, i]) => generic_1.addProof(tx, waves_crypto_1.signBytes(bytes1, s), i));
    tx.id = waves_crypto_1.hashBytes(bytes1);
    return tx;
}
exports.data = data;
