// waves-transactions/generic.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
const waves_crypto_1 = require("waves-crypto");
function getSenderPublicKey(seed, params) {
    const { seed: s } = exports.pullSeedAndIndex(seed);
    if (s == null && params.senderPublicKey == null)
        throw new Error('Please provide either seed or senderPublicKey');
    else {
        return params.senderPublicKey || waves_crypto_1.publicKey(s);
    }
}
exports.getSenderPublicKey = getSenderPublicKey;
exports.base64Prefix = (str) => str == null || str.slice(0, 7) === 'base64:' ? str : 'base64:' + str;
//export const isOne = <T>(oneOrMany: OneOrMany<T>): oneOrMany is T => !Array.isArray(oneOrMany)
//export const toMany = <T>(oneOrMany: OneOrMany<T>): T[] => isOne(oneOrMany) ? [oneOrMany] : oneOrMany
function addProof(tx, proof, index) {
    if (index == null) {
        tx.proofs = [...tx.proofs, proof];
        return tx;
    }
    if (tx.proofs != null && !!tx.proofs[index])
        throw new Error(`Proof at index ${index} is already exists.`);
    for (let i = tx.proofs.length; i < index; i++)
        tx.proofs.push('');
    tx.proofs[index] = proof;
    return tx;
}
exports.addProof = addProof;
exports.valOrDef = (val, def) => val != null ? val : def;
exports.isSeedsAndIndexes = (seed) => typeof seed !== 'string' && typeof seed === 'object' && seed.length === undefined;
exports.isArrayOfSeeds = (seed) => typeof seed !== 'string' && typeof seed === 'object' && seed.length !== undefined;
exports.mapSeed = (seed, map) => {
    const { seed: _seed, index } = exports.pullSeedAndIndex(seed);
    if (_seed != null)
        return map(_seed, index);
    return undefined;
};
exports.pullSeedAndIndex = (seed) => {
    const empty = { seed: undefined, index: undefined, nextSeed: undefined };
    if (seed == null || seed === '')
        return empty;
    if (exports.isSeedsAndIndexes(seed)) {
        const keys = Object.keys(seed).map(k => parseInt(k)).filter(k => !isNaN(k));
        if (keys == null || keys.length === 0)
            return empty;
        const index = keys[0];
        const newSeed = Object.assign({}, seed);
        delete newSeed[index];
        return { seed: seed[keys[0]], index, nextSeed: Object.keys(newSeed).length > 0 ? newSeed : undefined };
    }
    else if (exports.isArrayOfSeeds(seed)) {
        return exports.pullSeedAndIndex(Object.entries(seed).filter(([k, v]) => !!v).reduce((acc, [k, v]) => (Object.assign({}, acc, { [k]: v })), {}));
    }
    return { seed: seed };
};
