// waves-transactions/general.js
"use strict";
var __awaiter = (this && this.__awaiter) || function (thisArg, _arguments, P, generator) {
    return new (P || (P = Promise))(function (resolve, reject) {
        function fulfilled(value) { try { step(generator.next(value)); } catch (e) { reject(e); } }
        function rejected(value) { try { step(generator["throw"](value)); } catch (e) { reject(e); } }
        function step(result) { result.done ? resolve(result.value) : new P(function (resolve) { resolve(result.value); }).then(fulfilled, rejected); }
        step((generator = generator.apply(thisArg, _arguments || [])).next());
    });
};
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
const axios_1 = __importDefault(require("axios"));
const transactions_1 = require("./transactions");
const issue_1 = require("./transactions/issue");
const transfer_1 = require("./transactions/transfer");
const reissue_1 = require("./transactions/reissue");
const burn_1 = require("./transactions/burn");
const lease_1 = require("./transactions/lease");
const cancel_lease_1 = require("./transactions/cancel-lease");
const data_1 = require("./transactions/data");
const mass_transfer_1 = require("./transactions/mass-transfer");
const alias_1 = require("./transactions/alias");
const set_script_1 = require("./transactions/set-script");
const order_1 = require("./transactions/order");
const txToJson_1 = require("./txToJson");
const set_asset_script_1 = require("./transactions/set-asset-script");
exports.txTypeMap = {
    [transactions_1.TransactionType.Issue]: { sign: (x, seed) => issue_1.issue(x, seed), serialize: (x) => issue_1.issueToBytes(x) },
    [transactions_1.TransactionType.Transfer]: { sign: (x, seed) => transfer_1.transfer(x, seed), serialize: (x) => transfer_1.transferToBytes(x) },
    [transactions_1.TransactionType.Reissue]: { sign: (x, seed) => reissue_1.reissue(x, seed), serialize: (x) => reissue_1.reissueToBytes(x) },
    [transactions_1.TransactionType.Burn]: { sign: (x, seed) => burn_1.burn(x, seed), serialize: (x) => burn_1.burnToBytes(x) },
    [transactions_1.TransactionType.Lease]: { sign: (x, seed) => lease_1.lease(x, seed), serialize: (x) => lease_1.leaseToBytes(x) },
    [transactions_1.TransactionType.CancelLease]: { sign: (x, seed) => cancel_lease_1.cancelLease(x, seed), serialize: (x) => cancel_lease_1.cancelLeaseToBytes(x) },
    [transactions_1.TransactionType.Alias]: { sign: (x, seed) => alias_1.alias(x, seed), serialize: (x) => alias_1.aliasToBytes(x) },
    [transactions_1.TransactionType.MassTransfer]: { sign: (x, seed) => mass_transfer_1.massTransfer(x, seed), serialize: (x) => mass_transfer_1.massTransferToBytes(x) },
    [transactions_1.TransactionType.Data]: { sign: (x, seed) => data_1.data(x, seed), serialize: (x) => data_1.dataToBytes(x) },
    [transactions_1.TransactionType.SetScript]: { sign: (x, seed) => set_script_1.setScript(x, seed), serialize: (x) => set_script_1.setScriptToBytes(x) },
    [transactions_1.TransactionType.SetAssetScript]: { sign: (x, seed) => set_asset_script_1.setAssetScript(x, seed), serialize: (x) => set_asset_script_1.setAssetScriptToBytes(x) },
};
exports.signTx = (tx, seed) => {
    if (!exports.txTypeMap[tx.type])
        throw new Error(`Unknown tx type: ${tx.type}`);
    return exports.txTypeMap[tx.type].sign(tx, seed);
};
exports.serialize = (obj) => {
    if (order_1.isOrder(obj))
        return order_1.orderToBytes(obj);
    if (!exports.txTypeMap[obj.type])
        throw new Error(`Unknown tx type: ${obj.type}`);
    return exports.txTypeMap[obj.type].serialize(obj);
};
exports.broadcast = (tx, apiBase) => axios_1.default.post('transactions/broadcast', txToJson_1.txToJson(tx), { baseURL: apiBase, headers: { 'content-type': 'application/json' } })
    .then(x => x.data)
    .catch(e => Promise.reject(e.response && e.response.status === 400 ? new Error(e.response.data.message) : e));
exports.delay = (timeout) => {
    const t = {};
    const p = new Promise((resolve, _) => {
        t.resolve = resolve;
        t.id = setTimeout(() => resolve(), timeout);
    });
    p.cancel = () => { t.resolve(); clearTimeout(t.id); };
    return p;
};
exports.waitForTx = (txId, timeout, apiBase) => __awaiter(this, void 0, void 0, function* () {
    const promise = () => axios_1.default.get(`transactions/info/${txId}`, { baseURL: apiBase })
        .then(x => x.data).catch(_ => exports.delay(1000).then(_ => promise()));
    const t = exports.delay(timeout);
    const r = yield Promise.race([t.then(x => Promise.reject('timeout')), promise()]);
    t.cancel();
    return r;
});
