// waves-transactions/transactions.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
var TransactionType;
(function (TransactionType) {
    TransactionType[TransactionType["Genesis"] = 1] = "Genesis";
    TransactionType[TransactionType["Payment"] = 2] = "Payment";
    TransactionType[TransactionType["Issue"] = 3] = "Issue";
    TransactionType[TransactionType["Transfer"] = 4] = "Transfer";
    TransactionType[TransactionType["Reissue"] = 5] = "Reissue";
    TransactionType[TransactionType["Burn"] = 6] = "Burn";
    TransactionType[TransactionType["Exchange"] = 7] = "Exchange";
    TransactionType[TransactionType["Lease"] = 8] = "Lease";
    TransactionType[TransactionType["CancelLease"] = 9] = "CancelLease";
    TransactionType[TransactionType["Alias"] = 10] = "Alias";
    TransactionType[TransactionType["MassTransfer"] = 11] = "MassTransfer";
    TransactionType[TransactionType["Data"] = 12] = "Data";
    TransactionType[TransactionType["SetScript"] = 13] = "SetScript";
    TransactionType[TransactionType["SponsorFee"] = 14] = "SponsorFee";
    TransactionType[TransactionType["SetAssetScript"] = 15] = "SetAssetScript";
})(TransactionType = exports.TransactionType || (exports.TransactionType = {}));
