// @waves/signature-adapter/dist/prepareTx/constants.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
var _a;
var signature_generator_1 = require("@waves/signature-generator");
var TRANSACTION_TYPE_NUMBER;
(function (TRANSACTION_TYPE_NUMBER) {
    TRANSACTION_TYPE_NUMBER[TRANSACTION_TYPE_NUMBER["SEND_OLD"] = 2] = "SEND_OLD";
    TRANSACTION_TYPE_NUMBER[TRANSACTION_TYPE_NUMBER["ISSUE"] = 3] = "ISSUE";
    TRANSACTION_TYPE_NUMBER[TRANSACTION_TYPE_NUMBER["TRANSFER"] = 4] = "TRANSFER";
    TRANSACTION_TYPE_NUMBER[TRANSACTION_TYPE_NUMBER["REISSUE"] = 5] = "REISSUE";
    TRANSACTION_TYPE_NUMBER[TRANSACTION_TYPE_NUMBER["BURN"] = 6] = "BURN";
    TRANSACTION_TYPE_NUMBER[TRANSACTION_TYPE_NUMBER["EXCHANGE"] = 7] = "EXCHANGE";
    TRANSACTION_TYPE_NUMBER[TRANSACTION_TYPE_NUMBER["LEASE"] = 8] = "LEASE";
    TRANSACTION_TYPE_NUMBER[TRANSACTION_TYPE_NUMBER["CANCEL_LEASING"] = 9] = "CANCEL_LEASING";
    TRANSACTION_TYPE_NUMBER[TRANSACTION_TYPE_NUMBER["CREATE_ALIAS"] = 10] = "CREATE_ALIAS";
    TRANSACTION_TYPE_NUMBER[TRANSACTION_TYPE_NUMBER["MASS_TRANSFER"] = 11] = "MASS_TRANSFER";
    TRANSACTION_TYPE_NUMBER[TRANSACTION_TYPE_NUMBER["DATA"] = 12] = "DATA";
    TRANSACTION_TYPE_NUMBER[TRANSACTION_TYPE_NUMBER["SET_SCRIPT"] = 13] = "SET_SCRIPT";
    TRANSACTION_TYPE_NUMBER[TRANSACTION_TYPE_NUMBER["SPONSORSHIP"] = 14] = "SPONSORSHIP";
})(TRANSACTION_TYPE_NUMBER = exports.TRANSACTION_TYPE_NUMBER || (exports.TRANSACTION_TYPE_NUMBER = {}));
var SIGN_TYPE;
(function (SIGN_TYPE) {
    SIGN_TYPE[SIGN_TYPE["AUTH"] = 1000] = "AUTH";
    SIGN_TYPE[SIGN_TYPE["MATCHER_ORDERS"] = 1001] = "MATCHER_ORDERS";
    SIGN_TYPE[SIGN_TYPE["CREATE_ORDER"] = 1002] = "CREATE_ORDER";
    SIGN_TYPE[SIGN_TYPE["CANCEL_ORDER"] = 1003] = "CANCEL_ORDER";
    SIGN_TYPE[SIGN_TYPE["COINOMAT_CONFIRMATION"] = 1004] = "COINOMAT_CONFIRMATION";
    SIGN_TYPE[SIGN_TYPE["CUSTOM"] = 1005] = "CUSTOM";
    SIGN_TYPE[SIGN_TYPE["ISSUE"] = 3] = "ISSUE";
    SIGN_TYPE[SIGN_TYPE["TRANSFER"] = 4] = "TRANSFER";
    SIGN_TYPE[SIGN_TYPE["REISSUE"] = 5] = "REISSUE";
    SIGN_TYPE[SIGN_TYPE["BURN"] = 6] = "BURN";
    SIGN_TYPE[SIGN_TYPE["EXCHANGE"] = 7] = "EXCHANGE";
    SIGN_TYPE[SIGN_TYPE["LEASE"] = 8] = "LEASE";
    SIGN_TYPE[SIGN_TYPE["CANCEL_LEASING"] = 9] = "CANCEL_LEASING";
    SIGN_TYPE[SIGN_TYPE["CREATE_ALIAS"] = 10] = "CREATE_ALIAS";
    SIGN_TYPE[SIGN_TYPE["MASS_TRANSFER"] = 11] = "MASS_TRANSFER";
    SIGN_TYPE[SIGN_TYPE["DATA"] = 12] = "DATA";
    SIGN_TYPE[SIGN_TYPE["SET_SCRIPT"] = 13] = "SET_SCRIPT";
    SIGN_TYPE[SIGN_TYPE["SPONSORSHIP"] = 14] = "SPONSORSHIP";
})(SIGN_TYPE = exports.SIGN_TYPE || (exports.SIGN_TYPE = {}));
exports.SIGN_TYPES = (_a = {},
    _a[SIGN_TYPE.AUTH] = {
        signatureGenerator: signature_generator_1.generate([
            new signature_generator_1.StringWithLength('prefix'),
            new signature_generator_1.StringWithLength('host'),
            new signature_generator_1.StringWithLength('data')
        ]),
        adapter: 'signRequest'
    },
    _a[SIGN_TYPE.COINOMAT_CONFIRMATION] = {
        signatureGenerator: signature_generator_1.generate([
            new signature_generator_1.StringWithLength('prefix'),
            new signature_generator_1.Long('timestamp')
        ]),
        adapter: 'signRequest'
    },
    _a[SIGN_TYPE.MATCHER_ORDERS] = {
        signatureGenerator: signature_generator_1.AUTH_ORDER_SIGNATURE,
        adapter: 'signRequest'
    },
    _a[SIGN_TYPE.CREATE_ORDER] = {
        signatureGenerator: signature_generator_1.CREATE_ORDER_SIGNATURE,
        adapter: 'signOrder'
    },
    _a[SIGN_TYPE.CANCEL_ORDER] = {
        signatureGenerator: signature_generator_1.CANCEL_ORDER_SIGNATURE,
        adapter: 'signRequest'
    },
    _a[SIGN_TYPE.TRANSFER] = {
        signatureGenerator: signature_generator_1.TX_NUMBER_MAP[TRANSACTION_TYPE_NUMBER.TRANSFER],
        adapter: 'signTransaction'
    },
    _a[SIGN_TYPE.ISSUE] = {
        signatureGenerator: signature_generator_1.TX_NUMBER_MAP[TRANSACTION_TYPE_NUMBER.ISSUE],
        adapter: 'signTransaction'
    },
    _a[SIGN_TYPE.REISSUE] = {
        signatureGenerator: signature_generator_1.TX_NUMBER_MAP[TRANSACTION_TYPE_NUMBER.REISSUE],
        adapter: 'signTransaction'
    },
    _a[SIGN_TYPE.BURN] = {
        signatureGenerator: signature_generator_1.TX_NUMBER_MAP[TRANSACTION_TYPE_NUMBER.BURN],
        adapter: 'signTransaction'
    },
    _a[SIGN_TYPE.EXCHANGE] = {
        signatureGenerator: signature_generator_1.generate([
            new signature_generator_1.StringWithLength('exchange-tx')
        ]),
        adapter: 'signRequest'
    },
    _a[SIGN_TYPE.LEASE] = {
        signatureGenerator: signature_generator_1.TX_NUMBER_MAP[TRANSACTION_TYPE_NUMBER.LEASE],
        adapter: 'signTransaction'
    },
    _a[SIGN_TYPE.CANCEL_LEASING] = {
        signatureGenerator: signature_generator_1.TX_NUMBER_MAP[TRANSACTION_TYPE_NUMBER.CANCEL_LEASING],
        adapter: 'signTransaction'
    },
    _a[SIGN_TYPE.CREATE_ALIAS] = {
        signatureGenerator: signature_generator_1.TX_NUMBER_MAP[TRANSACTION_TYPE_NUMBER.CREATE_ALIAS],
        adapter: 'signTransaction'
    },
    _a[SIGN_TYPE.MASS_TRANSFER] = {
        signatureGenerator: signature_generator_1.TX_NUMBER_MAP[TRANSACTION_TYPE_NUMBER.MASS_TRANSFER],
        adapter: 'signTransaction'
    },
    _a[SIGN_TYPE.DATA] = {
        signatureGenerator: signature_generator_1.TX_NUMBER_MAP[TRANSACTION_TYPE_NUMBER.DATA],
        adapter: 'signTransaction'
    },
    _a[SIGN_TYPE.SET_SCRIPT] = {
        signatureGenerator: signature_generator_1.TX_NUMBER_MAP[TRANSACTION_TYPE_NUMBER.SET_SCRIPT],
        adapter: 'signTransaction'
    },
    _a[SIGN_TYPE.SPONSORSHIP] = {
        signatureGenerator: signature_generator_1.TX_NUMBER_MAP[TRANSACTION_TYPE_NUMBER.SPONSORSHIP],
        adapter: 'signTransaction'
    },
    _a);
