// @waves/signature-adapter/dist/prepareTx/fieldTypes.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
var fieldFactory = function (type) { return function (fromField, toField, processor, optional) {
    if (toField === void 0) { toField = fromField; }
    if (processor === void 0) { processor = null; }
    if (optional === void 0) { optional = false; }
    return ({
        name: fromField,
        field: toField || fromField,
        optional: optional,
        processor: processor,
        type: type
    });
}; };
exports.string = fieldFactory('string');
exports.asset = fieldFactory('assetId');
exports.assetName = fieldFactory('assetName');
exports.assetDescription = fieldFactory('assetDescription');
exports.precision = fieldFactory('precision');
exports.number = fieldFactory('number');
exports.address = fieldFactory('address');
exports.aliasName = fieldFactory('aliasName');
exports.aliasOrAddress = fieldFactory('aliasOrAddress');
exports.money = fieldFactory('money');
exports.numberLike = fieldFactory('numberLike');
exports.attachment = fieldFactory('attachment');
exports.httpsUrl = fieldFactory('httpsUrl');
exports.timestamp = fieldFactory('timestamp');
exports.orderType = fieldFactory('orderType');
exports.fromData = fieldFactory('fromData');
exports.boolean = fieldFactory('boolean');
exports.transfers = fieldFactory('transfers');
exports.data = fieldFactory('data');
exports.script = fieldFactory('script');
exports.required = fieldFactory('required');
