// @waves/signature-adapter/dist/adapters/MetaMaskAdapter.js
"use strict";
var __extends = (this && this.__extends) || (function () {
    var extendStatics = Object.setPrototypeOf ||
        ({ __proto__: [] } instanceof Array && function (d, b) { d.__proto__ = b; }) ||
        function (d, b) { for (var p in b) if (b.hasOwnProperty(p)) d[p] = b[p]; };
    return function (d, b) {
        extendStatics(d, b);
        function __() { this.constructor = d; }
        d.prototype = b === null ? Object.create(b) : (__.prototype = b.prototype, new __());
    };
})();
Object.defineProperty(exports, "__esModule", { value: true });
var Adapter_1 = require("./Adapter");
var MetaMaskAdapter = /** @class */ (function (_super) {
    __extends(MetaMaskAdapter, _super);
    function MetaMaskAdapter() {
        return _super !== null && _super.apply(this, arguments) || this;
    }
    MetaMaskAdapter.prototype.getPublicKey = function () {
        return Promise.resolve(''); // TODO
    };
    MetaMaskAdapter.prototype.getAddress = function () {
        return Promise.resolve(''); // TODO
    };
    MetaMaskAdapter.prototype.getSeed = function () {
        return Promise.reject(Error('Method "getSeed" is not available!'));
    };
    MetaMaskAdapter.prototype.signRequest = function (bytes) {
        return Promise.resolve(''); //TODO
    };
    MetaMaskAdapter.prototype.signTransaction = function (bytes, amountPrecision) {
        return Promise.resolve(''); //TODO
    };
    MetaMaskAdapter.prototype.signOrder = function (bytes, amountPrecision) {
        return Promise.resolve(''); //TODO
    };
    MetaMaskAdapter.prototype.signData = function (bytes) {
        return Promise.resolve(''); //TODO
    };
    MetaMaskAdapter.prototype.getPrivateKey = function () {
        return Promise.reject('No private key');
    };
    MetaMaskAdapter.isAvailable = function () {
        return Promise.resolve(false);
    };
    MetaMaskAdapter.type = "metaMask" /* MetaMask */;
    return MetaMaskAdapter;
}(Adapter_1.Adapter));
exports.MetaMaskAdapter = MetaMaskAdapter;
