// @waves/signature-adapter/dist/adapters/WavesKeeperAdapter.js
"use strict";
var __extends = (this && this.__extends) || (function () {
    var extendStatics = Object.setPrototypeOf ||
        ({ __proto__: [] } instanceof Array && function (d, b) { d.__proto__ = b; }) ||
        function (d, b) { for (var p in b) if (b.hasOwnProperty(p)) d[p] = b[p]; };
    return function (d, b) {
        extendStatics(d, b);
        function __() { this.constructor = d; }
        d.prototype = b === null ? Object.create(b) : (__.prototype = b.prototype, new __());
    };
})();
var __awaiter = (this && this.__awaiter) || function (thisArg, _arguments, P, generator) {
    return new (P || (P = Promise))(function (resolve, reject) {
        function fulfilled(value) { try { step(generator.next(value)); } catch (e) { reject(e); } }
        function rejected(value) { try { step(generator["throw"](value)); } catch (e) { reject(e); } }
        function step(result) { result.done ? resolve(result.value) : new P(function (resolve) { resolve(result.value); }).then(fulfilled, rejected); }
        step((generator = generator.apply(thisArg, _arguments || [])).next());
    });
};
var __generator = (this && this.__generator) || function (thisArg, body) {
    var _ = { label: 0, sent: function() { if (t[0] & 1) throw t[1]; return t[1]; }, trys: [], ops: [] }, f, y, t, g;
    return g = { next: verb(0), "throw": verb(1), "return": verb(2) }, typeof Symbol === "function" && (g[Symbol.iterator] = function() { return this; }), g;
    function verb(n) { return function (v) { return step([n, v]); }; }
    function step(op) {
        if (f) throw new TypeError("Generator is already executing.");
        while (_) try {
            if (f = 1, y && (t = op[0] & 2 ? y["return"] : op[0] ? y["throw"] || ((t = y["return"]) && t.call(y), 0) : y.next) && !(t = t.call(y, op[1])).done) return t;
            if (y = 0, t) op = [op[0] & 2, t.value];
            switch (op[0]) {
                case 0: case 1: t = op; break;
                case 4: _.label++; return { value: op[1], done: false };
                case 5: _.label++; y = op[1]; op = [0]; continue;
                case 7: op = _.ops.pop(); _.trys.pop(); continue;
                default:
                    if (!(t = _.trys, t = t.length > 0 && t[t.length - 1]) && (op[0] === 6 || op[0] === 2)) { _ = 0; continue; }
                    if (op[0] === 3 && (!t || (op[1] > t[0] && op[1] < t[3]))) { _.label = op[1]; break; }
                    if (op[0] === 6 && _.label < t[1]) { _.label = t[1]; t = op; break; }
                    if (t && _.label < t[2]) { _.label = t[2]; _.ops.push(op); break; }
                    if (t[2]) _.ops.pop();
                    _.trys.pop(); continue;
            }
            op = body.call(thisArg, _);
        } catch (e) { op = [6, e]; y = 0; } finally { f = t = 0; }
        if (op[0] & 5) throw op[1]; return { value: op[0] ? op[1] : void 0, done: true };
    }
};
Object.defineProperty(exports, "__esModule", { value: true });
var Adapter_1 = require("./Adapter");
var prepareTx_1 = require("../prepareTx");
var signature_generator_1 = require("@waves/signature-generator");
var WavesKeeperAdapter = /** @class */ (function (_super) {
    __extends(WavesKeeperAdapter, _super);
    function WavesKeeperAdapter(_a) {
        var address = _a.address, publicKey = _a.publicKey;
        var _this = _super.call(this) || this;
        _this._onDestoryCb = [];
        _this._needDestroy = false;
        WavesKeeperAdapter._initExtension();
        _this._address = address;
        _this._pKey = publicKey;
        WavesKeeperAdapter.onUpdate(function (state) {
            if (!state.locked && (!state.account || state.account.address !== _this._address)) {
                _this._needDestroy = true;
                _this._onDestoryCb.forEach(function (cb) { return cb(); });
            }
        });
        return _this;
    }
    WavesKeeperAdapter.prototype.isAvailable = function (ignoreLocked) {
        if (ignoreLocked === void 0) { ignoreLocked = false; }
        return __awaiter(this, void 0, void 0, function () {
            var data, e_1;
            return __generator(this, function (_a) {
                switch (_a.label) {
                    case 0:
                        _a.trys.push([0, 3, , 4]);
                        return [4 /*yield*/, WavesKeeperAdapter.isAvailable()];
                    case 1:
                        _a.sent();
                        return [4 /*yield*/, WavesKeeperAdapter._api.publicState()];
                    case 2:
                        data = _a.sent();
                        if (data.locked) {
                            return [2 /*return*/, ignoreLocked ? Promise.resolve() : Promise.reject({ code: 4, msg: 'Keeper is locked' })];
                        }
                        if (data.account && data.account.address === this._address) {
                            return [2 /*return*/, Promise.resolve()];
                        }
                        return [3 /*break*/, 4];
                    case 3:
                        e_1 = _a.sent();
                        return [3 /*break*/, 4];
                    case 4: return [2 /*return*/, Promise.reject({ code: 5, msg: 'Keeper has another active account' })];
                }
            });
        });
    };
    WavesKeeperAdapter.prototype.isLocked = function () {
        return __awaiter(this, void 0, void 0, function () {
            var data;
            return __generator(this, function (_a) {
                switch (_a.label) {
                    case 0: return [4 /*yield*/, WavesKeeperAdapter.isAvailable()];
                    case 1:
                        _a.sent();
                        return [4 /*yield*/, WavesKeeperAdapter._api.publicState()];
                    case 2:
                        data = _a.sent();
                        if (data.locked) {
                            return [2 /*return*/, Promise.resolve()];
                        }
                        return [2 /*return*/];
                }
            });
        });
    };
    WavesKeeperAdapter.prototype.onDestroy = function (cb) {
        if (this._needDestroy) {
            return cb();
        }
        this._onDestoryCb.push(cb);
    };
    WavesKeeperAdapter.prototype.getPublicKey = function () {
        return Promise.resolve(this._pKey);
    };
    WavesKeeperAdapter.prototype.getAddress = function () {
        return Promise.resolve(this._address);
    };
    WavesKeeperAdapter.prototype.getSeed = function () {
        return Promise.reject(Error('Method "getSeed" is not available!'));
    };
    WavesKeeperAdapter.prototype.signRequest = function (bytes, _, signData) {
        return __awaiter(this, void 0, void 0, function () {
            return __generator(this, function (_a) {
                switch (_a.label) {
                    case 0: return [4 /*yield*/, this.isAvailable(true)];
                    case 1:
                        _a.sent();
                        return [4 /*yield*/, WavesKeeperAdapter._api.signRequest(signData)];
                    case 2: return [2 /*return*/, _a.sent()];
                }
            });
        });
    };
    WavesKeeperAdapter.prototype.signTransaction = function (bytes, amountPrecision, signData) {
        return __awaiter(this, void 0, void 0, function () {
            var dataStr, _a, proofs, signature;
            return __generator(this, function (_b) {
                switch (_b.label) {
                    case 0: return [4 /*yield*/, this.isAvailable(true)];
                    case 1:
                        _b.sent();
                        return [4 /*yield*/, WavesKeeperAdapter._api.signTransaction(signData)];
                    case 2:
                        dataStr = _b.sent();
                        _a = JSON.parse(dataStr), proofs = _a.proofs, signature = _a.signature;
                        return [2 /*return*/, signature || proofs.pop()];
                }
            });
        });
    };
    WavesKeeperAdapter.prototype.signOrder = function (bytes, amountPrecision, signData) {
        return __awaiter(this, void 0, void 0, function () {
            var promise, dataStr, _a, proofs, signature;
            return __generator(this, function (_b) {
                switch (_b.label) {
                    case 0: return [4 /*yield*/, this.isAvailable(true)];
                    case 1:
                        _b.sent();
                        switch (signData.type) {
                            case prepareTx_1.SIGN_TYPE.CREATE_ORDER:
                                promise = WavesKeeperAdapter._api.signOrder(signData);
                                break;
                            case prepareTx_1.SIGN_TYPE.CANCEL_ORDER:
                                promise = WavesKeeperAdapter._api.signCancelOrder(signData);
                                break;
                            default:
                                return [2 /*return*/, WavesKeeperAdapter._api.signRequest(signData)];
                        }
                        return [4 /*yield*/, promise];
                    case 2:
                        dataStr = _b.sent();
                        _a = JSON.parse(dataStr), proofs = _a.proofs, signature = _a.signature;
                        return [2 /*return*/, signature || proofs.pop()];
                }
            });
        });
    };
    WavesKeeperAdapter.prototype.signData = function (bytes) {
        return __awaiter(this, void 0, void 0, function () {
            return __generator(this, function (_a) {
                switch (_a.label) {
                    case 0: return [4 /*yield*/, this.isAvailable(true)];
                    case 1:
                        _a.sent();
                        return [2 /*return*/, Promise.resolve('')]; //TODO
                }
            });
        });
    };
    WavesKeeperAdapter.prototype.getPrivateKey = function () {
        return Promise.reject('No private key');
    };
    WavesKeeperAdapter.isAvailable = function () {
        return __awaiter(this, void 0, void 0, function () {
            var error, data, e_2;
            return __generator(this, function (_a) {
                switch (_a.label) {
                    case 0:
                        WavesKeeperAdapter._initExtension();
                        if (!this._api) {
                            throw { code: 0, message: 'Install WavesKeeper' };
                        }
                        _a.label = 1;
                    case 1:
                        _a.trys.push([1, 3, , 4]);
                        return [4 /*yield*/, this._api.publicState()];
                    case 2:
                        data = _a.sent();
                        return [3 /*break*/, 4];
                    case 3:
                        e_2 = _a.sent();
                        error = { code: 1, message: 'No permissions' };
                        return [3 /*break*/, 4];
                    case 4:
                        if (!error && data) {
                            if (!data.locked && !data.account) {
                                error = { code: 2, message: 'No accounts in waveskeeper' };
                            }
                            else if (!data.locked && (!data.account.address || !signature_generator_1.utils.crypto.isValidAddress(data.account.address))) {
                                error = { code: 3, message: 'Selected network incorrect' };
                            }
                        }
                        if (error) {
                            throw error;
                        }
                        return [2 /*return*/, true];
                }
            });
        });
    };
    WavesKeeperAdapter.getUserList = function () {
        return __awaiter(this, void 0, void 0, function () {
            return __generator(this, function (_a) {
                switch (_a.label) {
                    case 0: return [4 /*yield*/, WavesKeeperAdapter.isAvailable()];
                    case 1:
                        _a.sent();
                        return [2 /*return*/, WavesKeeperAdapter._api.publicState().then(function (_a) {
                                var account = _a.account;
                                return [account];
                            })];
                }
            });
        });
    };
    WavesKeeperAdapter.initOptions = function (options) {
        Adapter_1.Adapter.initOptions(options);
        this.setApiExtension(options.extension);
    };
    WavesKeeperAdapter.setApiExtension = function (extension) {
        var extensionCb;
        if (typeof extension === 'function') {
            extensionCb = extension;
        }
        else if (extension) {
            extensionCb = function () { return extension; };
        }
        WavesKeeperAdapter._getApiCb = extensionCb;
    };
    WavesKeeperAdapter.onUpdate = function (cb) {
        WavesKeeperAdapter._onUpdateCb.push(cb);
    };
    WavesKeeperAdapter._initExtension = function () {
        if (WavesKeeperAdapter._api || !WavesKeeperAdapter._getApiCb) {
            return null;
        }
        this._api = WavesKeeperAdapter._getApiCb();
        if (this._api) {
            this._api.on('update', function (state) {
                for (var _i = 0, _a = WavesKeeperAdapter._onUpdateCb; _i < _a.length; _i++) {
                    var cb = _a[_i];
                    cb(state);
                }
            });
        }
    };
    WavesKeeperAdapter.type = "wavesKeeper" /* WavesKeeper */;
    WavesKeeperAdapter._onUpdateCb = [];
    return WavesKeeperAdapter;
}(Adapter_1.Adapter));
exports.WavesKeeperAdapter = WavesKeeperAdapter;
